estimagic/logging/create_database.py

```diff
@@ -0,0 +1,238 @@
+    """Return database metadata object for the database stored in ``path``.
+
+    This is the default way of loading a database for read-only purposes in estimagic.
+
+    Args:
+        path (str or pathlib.Path): location of the database file. If the file does
+            not exist, it will be created.
+        replace (bool): If true and the database exists, it will be overwritten.
+            Otherwise, data will be appended.
+
+    Returns:
+        metadata (sqlalchemy.MetaData). The engine that connects to the database can be
+        accessed via ``metadata.bind``.
+
+    """
+    if isinstance(path, str):
+        path = Path(path)
+
+    if path.exists() and replace:
+        path.unlink()
+
+    engine = create_engine(f"sqlite:///{path}")
+    _make_engine_thread_safe(engine)
+    metadata = MetaData()
+    metadata.bind = engine
+    metadata.reflect()
+    return metadata
+
+
+def _make_engine_thread_safe(engine):
+    """Make the engine even more thread safe than by default.
+
+    The code is taken from the documentation: https://tinyurl.com/u9xea5z
+
+    The main purpose is to emit the begin statement of any connection
+    as late as possible in order to keep the time in which the database
+    is locked as short as possible.
+
+    """
+
+    @event.listens_for(engine, "connect")
+    def do_connect(dbapi_connection, connection_record):
+        # disable pysqlite's emitting of the BEGIN statement entirely.
+        # also stops it from emitting COMMIT before absolutely necessary.
+        dbapi_connection.isolation_level = None
+
+    @event.listens_for(engine, "begin")
+    def do_begin(conn):
+        # emit our own BEGIN
+        conn.execute("BEGIN DEFERRED")
+
+
+def prepare_database(
+    path,
+    params,
+    db_options=None,
+    optimization_status="scheduled",
+    gradient_status=0,
+    replace=False,
+):
+    """Return database metadata object with all relevant tables.
+
+    This should always be used to create entirely new databases or to create the
+    tables needed during optimization in an existing database.
+
+    The resulting database has the following tables:
+    - params_history: the complete history of parameters from the optimization. The
+      first column is "iteration", the remaining columns are parameter names taken
+      from params["name"].
+    - gradient_history: the complete history of gradient evaluations from the
+      optimization. Same columns as params_history.
+    - criterion_history: the complete history of criterion values from the optimization.
+      The first column is "iteration", the second column is "value".
+    - time_stamps: timestamps from the end of each criterion evaluation. Same columss as
+      criterion_history.
+    - convergence_history: the complete history of convergence criteria from the
+      optimization. The first column is iteration, the other columns are "ftol", "gtol"
+      and "xtol".
+    - start_params: copy of user provided params.
```

Comment:
Is this not always the first row of `params_history`? A utility function to get that back would seem more helpful than storing the same thing twice.

Reply:
No, it is not. "start_params" contains the full params DataFrame with bounds columns, the correct Index (whereas the log uses parameter names as index to avoid a MultiIndex), information on fixed parameters, ... 

I made it clearer in the docstring.